File: src/analytics-service/app/services/monte_carlo.py
import numpy as np

N_SIMS = 10_000

# Historical annual parameters for Indian asset classes
_PARAMS = {
    "equity": {"mean": 0.12, "std": 0.18},
    "debt":   {"mean": 0.07, "std": 0.04},
    "gold":   {"mean": 0.08, "std": 0.15},
    "cash":   {"mean": 0.04, "std": 0.005},
}


def simulate_goal(
    target_amount: float,
    years_to_goal: float,
    current_value: float,
    monthly_contribution: float,
    equity_pct: float,
    debt_pct: float,
    gold_pct: float,
    cash_pct: float,
    n_sims: int = N_SIMS,
    seed: int | None = None,
) -> dict:
    """Run Monte Carlo simulation and return probability + corpus percentiles."""
    rng = np.random.default_rng(seed)
    months = max(1, int(round(years_to_goal * 12)))

    weights = np.array([equity_pct, debt_pct, gold_pct, cash_pct]) / 100.0
    annual_means = np.array([
        _PARAMS["equity"]["mean"], _PARAMS["debt"]["mean"],
        _PARAMS["gold"]["mean"],   _PARAMS["cash"]["mean"],
    ])
    annual_stds = np.array([
        _PARAMS["equity"]["std"], _PARAMS["debt"]["std"],
        _PARAMS["gold"]["std"],   _PARAMS["cash"]["std"],
    ])

    # Convert to monthly parameters
    monthly_means = annual_means / 12
    monthly_stds = annual_stds / np.sqrt(12)

    # Blended portfolio monthly parameters (simplified — no correlation matrix)
    port_mean = float(np.dot(weights, monthly_means))
    port_std = float(np.sqrt(np.dot(weights**2, monthly_stds**2)))

    # Generate all monthly returns at once: shape (n_sims, months)
    monthly_returns = rng.normal(port_mean, port_std, size=(n_sims, months))

    # Simulate corpus month by month (vectorised across all paths)
    corpus = np.full(n_sims, float(current_value))
    for m in range(months):
        corpus = corpus * (1.0 + monthly_returns[:, m]) + float(monthly_contribution)
    corpus = np.maximum(corpus, 0.0)

    probability = float(np.mean(corpus >= float(target_amount)))

    return {
        "probability_of_success": round(probability * 100.0, 1),
        "p10_corpus": round(float(np.percentile(corpus, 10)), 2),
        "p50_corpus": round(float(np.percentile(corpus, 50)), 2),
        "p90_corpus": round(float(np.percentile(corpus, 90)), 2),
    }

File: mobile/src/lib/finance/montecarlo.ts
// Monte Carlo goal simulation — ported from analytics-service/app/services/monte_carlo.py.
// numpy vectorisation replaced with plain loops; n_sims reduced to 2000 for on-device speed.
// Two-factor correlation model: r_i = sqrt(rho)*Z_market + sqrt(1-rho)*Z_idio.

export interface AssetAllocation {
  equityPct: number;
  debtPct: number;
  goldPct: number;
  cashPct: number;
}

export interface SimulationResult {
  probabilityOfSuccess: number; // percent 0–100
  p10Corpus: number;
  p50Corpus: number;
  p90Corpus: number;
}

const PARAMS = {
  equity: { mean: 0.12, std: 0.18 },
  debt: { mean: 0.07, std: 0.04 },
  gold: { mean: 0.08, std: 0.15 },
  cash: { mean: 0.04, std: 0.005 },
};

const DEFAULT_SIMS = 2000;

// Box-Muller standard normal
function randn(): number {
  let u = 0;
  let v = 0;
  while (u === 0) u = Math.random();
  while (v === 0) v = Math.random();
  return Math.sqrt(-2 * Math.log(u)) * Math.cos(2 * Math.PI * v);
}

function percentile(sortedAsc: number[], p: number): number {
  if (sortedAsc.length === 0) return 0;
  const idx = (p / 100) * (sortedAsc.length - 1);
  const lo = Math.floor(idx);
  const hi = Math.ceil(idx);
  if (lo === hi) return sortedAsc[lo];
  return sortedAsc[lo] + (sortedAsc[hi] - sortedAsc[lo]) * (idx - lo);
}

export function simulateGoal(
  targetAmount: number,
  yearsToGoal: number,
  currentValue: number,
  monthlyContribution: number,
  alloc: AssetAllocation,
  marketCorrelation = 0.3,
  nSims = DEFAULT_SIMS,
): SimulationResult {
  const months = Math.max(1, Math.round(yearsToGoal * 12));
  const weights = [alloc.equityPct, alloc.debtPct, alloc.goldPct, alloc.cashPct].map(
    (w) => w / 100,
  );
  const monthlyMeans = [
    PARAMS.equity.mean,
    PARAMS.debt.mean,
    PARAMS.gold.mean,
    PARAMS.cash.mean,
  ].map((m) => m / 12);
  const monthlyStds = [
    PARAMS.equity.std,
    PARAMS.debt.std,
    PARAMS.gold.std,
    PARAMS.cash.std,
  ].map((s) => s / Math.sqrt(12));

  const rho = Math.min(1, Math.max(0, marketCorrelation));
  const sqrtRho = Math.sqrt(rho);
  const sqrtOneMinus = Math.sqrt(1 - rho);

  const finalCorpus = new Array<number>(nSims);

  for (let s = 0; s < nSims; s++) {
    let corpus = currentValue;
    for (let m = 0; m < months; m++) {
      const zMarket = randn();
      // blended portfolio return this month
      let blended = 0;
      for (let k = 0; k < 4; k++) {
        const zIdio = randn();
        const assetReturn =
          sqrtRho * zMarket * monthlyStds[k] +
          sqrtOneMinus * zIdio * monthlyStds[k] +
          monthlyMeans[k];
        blended += assetReturn * weights[k];
      }
      corpus = corpus * (1 + blended) + monthlyContribution;
    }
    finalCorpus[s] = Math.max(0, corpus);
  }

  const successes = finalCorpus.reduce((n, c) => n + (c >= targetAmount ? 1 : 0), 0);
  const sorted = [...finalCorpus].sort((a, b) => a - b);

  return {
    probabilityOfSuccess: Math.round((successes / nSims) * 1000) / 10,
    p10Corpus: Math.round(percentile(sorted, 10) * 100) / 100,
    p50Corpus: Math.round(percentile(sorted, 50) * 100) / 100,
    p90Corpus: Math.round(percentile(sorted, 90) * 100) / 100,
  };
}
